### PR TITLE
mgr/dashboard: Settings service

### PR DESCRIPTION
**Pull request by Devp00l**
The settings service handles settings that are set
within the dashboard. (Prometheus Alertmanager (pending PR) and Grafana)

The service has a helpful function and a good memory. The function will
allow to check if a certain option is set, if it's set the callback will
be triggered. If the API request was made before for this particular
option, no API request is made anymore, as the service will remember if
the state is set or not.

Fixes: https://tracker.ceph.com/issues/37471
Signed-off-by: Stephan Müller <smueller@suse.com>

- [x] References tracker ticket
- [x] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.ts [hunk 1]
```diff
@@ -64,15 +64,7 @@ export class GrafanaComponent implements OnInit, OnChanges {
...
-    this.grafanaService.getGrafanaApiUrl().subscribe((data: any) => {
-      this.grafanaUrl = data.instance;
-      if (this.grafanaUrl === '') {
-        this.grafanaExist = false;
-        return;
-      } else {
-        this.getFrame();
-      }
-    });
+    this.grafanaService.ifGrafanaConfigured(() => this.getFrame());
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/components/grafana/grafana.component.ts]: This is currently wrong, as it does not set the `grafanaUrl` anywhere.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/grafana.service.ts [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
 @Injectable({
   providedIn: ApiModule
 })
 export class GrafanaService {
-  constructor(private http: HttpClient) {}
+  constructor(private settingsService: SettingsService) {}
 
-  getGrafanaApiUrl() {
-    return this.http.get('api/grafana/url');
+  ifGrafanaConfigured(fn): void {
```
[the remark this answers is not in the document]
Comment (Devp00l) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/api/grafana.service.ts]: Not yet but it's needed in the prometheus PR ;) I used the settings service file there before - but than it got renamed :P

**Comment by Devp00l** [comment]
Addressed all comments

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/settings.service.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  constructor(private http: HttpClient) {}
+
+  private settings: { [url: string]: boolean } = {};
+
+  ifOptionConfigured(url: string, fn: Function): void {
+    const option = this.settings[url];
+    if (option === undefined) {
+      this.http.get(url).subscribe((data: any) => {
+        const isNotEmpty = (value) => _.isString(value) && value !== '';
+        this.settings[url] = isNotEmpty(data.value) || isNotEmpty(data.instance);
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/api/settings.service.ts]: And why not storing the setting value, instead of just the boolean?

**Comment by Devp00l** [comment]
Addressed all comments

**Comment by s0nea** [comment]
Hey @Devp00l, may I ask you to give a bit more context why this is needed? You added a description  about WHAT is covered by the PR but not WHY. Could you also tell if and how it's related to https://github.com/ceph/ceph/pull/22457 please?

**Pull request merged**